ingest/exporter/bundle_update_service.py

```diff
@@ -0,0 +1,175 @@
+from ingest.api.stagingapi import StagingApi, MetadataFileStagingRequest, FileDescription
+from ingest.api.dssapi import DssApi, DSSFileCreateRequest
+from ingest.api.ingestapi import IngestApi
+
+from typing import Iterable, Dict
+
+
+class MetadataToBeBundled:
+    def __init__(self, metadata_type, uuid, version, cloud_url, file_descriptor):
+        self.metadata_type = metadata_type
+        self.uuid = uuid
+        self.version = version
+        self.cloud_url = cloud_url
+        self.file_descriptor = file_descriptor
+
+
+class MetadataResource:
+    def __init__(self, metadata_resource, metadata_type, metadata_json, uuid, dcp_version):
+        self.metadata_resource = metadata_resource
+        self.metdata_type = metadata_type
+        self.metadata_json = metadata_json
+        self.uuid = uuid
+        self.dcp_version = dcp_version
+
+
+class StagedMetadataResource:
+    def __init__(self, metadata_resource: MetadataResource, staged_file_description: FileDescription):
+        self.metadata_resource = metadata_resource
+        self.staged_file_description = staged_file_description
+
+
+class DSSMetadataFileResource:
+    def __init__(self, dss_file_resource: dict, uuid: str, metadata_resource: MetadataResource):
+        self.dss_file_resource = dss_file_resource
+        self.uuid = uuid
+        self.metadata_resource = metadata_resource
+
+
+class BundleUpdateService:
+    def __init__(self, staging_client: StagingApi, dss_client: DssApi, ingest_client: IngestApi):
+        self.staging_client = staging_client
+        self.dss_client = dss_client
+        self.ingest_client = ingest_client
+
+    def update_bundle(self, update_submission: dict, bundle_uuid: str, updated_bundle_version: str, metadata_callbacks_to_update: Iterable[str]):
+        """
+        Note: tricky part here is updating bundles with the correct file-name within the bundle. To do that, we must know
+        what we named the file-to-be-updated and ensure to name its update with the same name.
+
+        :param update_submission:
+        :param bundles_to_update:
+        :param bundle_files_to_update:
+        :return:
+        """
+        staging_area_id = BundleUpdateService.upload_area_id_for_submission(update_submission)
+
+        # fetch the metadata_documents
+        metadata_resources = self.fetch_and_parse_metadata(metadata_callbacks_to_update)
+
+        # stage updates in the upload-service and store in dss
+        staged_metadata_resources = self.stage_metadata_resources(metadata_resources, staging_area_id)
```

Comment:
My big question here is how are we avoiding doing staging for lots of files that are shared between bundles (e.g. project_0.json)? It looks like we're receiving a bundle and the set of update files, which would mean staging again and again. But quite possibly I'm missing something. Perhaps something clever is going on upstream.

Reply:
Or maybe we're relying on the datastore to ignore these updates since they have the same version number as the existing file?